### PR TITLE
Style guide - code blocks

### PR DESCRIPTION
**Pull request by mongoKart**
# Pull Request Process

Please read the
[Style Guide Review Process](https://wiki.corp.mongodb.com/display/DE/Technical+Writer+Style+Guide+Review+Process)
wiki page.

Contributors should take the following actions:

- Request a PR review in #docs-style-guide Slack channel.
- Add the Style Guide team reps as reviewers.
- If any reviewer requests changes, address them and request another review.
- When you receive PR review approvals from all the Style Guide team reps,
  they will merge the PR and notify you of the merge in the Slack channel.

# Pull Request Description

No ticket--skunkworks project.

Staging URL: https://preview-mongodbmongokart.gatsbyjs.io/meta/style-guide-code-blocks/reference/code-blocks/

### CODE REVIEW

**Review by ccho-mongodb — CHANGES_REQUESTED** [review]
Thanks for this PR! I think the content and examples are great and achieves the learning goals we're looking for in the [Snooty RST Guide product brief](https://docs.google.com/document/d/1HRuMYSd-6Tzc0Hix7xnbaKkQAbzIaznCDYJBGkEI1RU/edit#heading=h.wdpdgdpcthgg).

However, I think it might belong in a new section rather than "Markup Guidelines", as that implies that it provides guidance (which I think is better to keep separate). I think the existing content in that section will need to be moved or restructured to match the type of guidance that "Writing Guidelines" and "Style Guidelines" pages provide.

Instead, perhaps this could be relocated to "Reference", perhaps replacing "Copyable Code". The current content in the Reference section matches more closely with the content in this PR. Let me know if you have any objections to that.

File: source/style-guide/markup/directives/code-blocks.txt [hunk 1]
```diff
@@ -2,20 +2,444 @@
...
+      :dedent: <number>
+
+Options
+~~~~~~~
+
+.. list-table::
+
+   * - ``<file path>``
+     - **Required.** The path to the file in the repository that contains the code you
+       want to show.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:335]:
```suggestion
     - **Required.** Path to the file in the repository that contains the code to show.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 2]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** None
+
+   * - ``:language: <language>``
+     - The language to use for syntax highlighting.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:340]:
```suggestion
     - Language to use for syntax highlighting. 
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 3]
```diff
@@ -2,20 +2,444 @@
...
+   * - ``:language: <language>``
+     - The language to use for syntax highlighting. 
+       
+       For a complete list of supported languages, see the
+       `leafygreen-ui GitHub repository. <https://github.com/mongodb/leafygreen-ui/tree/main/packages/code#properties>`__
+
+       **Default:** None
+ 
+   * - ``:copyable: <boolean>``
+     - Whether to show the Copy button for the code sample.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:348]:
```suggestion
     - Option that indicates whether to show the :guilabel:`Copy` button.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 4]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** ``true``
+
+   * - ``:start-after: <line of code>``
+     - The line of code before the first line that you want to include in the sample.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:353]:
```suggestion
     - Line of code before the first line to include in the sample.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 5]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** Beginning of file
+   
+   * - ``:end-before: <line of code>``
+     - The line of code after the last line that you want to include in the sample.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:358]:
```suggestion
     - Line of code after the last line to include in the sample.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 6]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** End of file
+
+   * - ``:linenos:``
+     - Shows line numbers for the code sample.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:363]:
```suggestion
     - Option that indicates whether to shows line numbers.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 7]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** None
+ 
+   * - ``:lineno-start:``
+     - The number at which to start line numbering. This option requires the
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:368]:
```suggestion
     - Number of the line at which to start showing line numbers. This option requires the
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 8]
```diff
@@ -2,20 +2,444 @@
...
+       ``:linenos:`` option.
+
+       **Default:** 1
+
+   * - ``:emphasize-lines: <line number>``
+     - Lines to highlight for the code sample. You can specify a range
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:374]:
```suggestion
     - Lines to highlight. You can specify a range 
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 9]
```diff
@@ -2,20 +2,444 @@
...
+       **Default:** 1
+
+   * - ``:emphasize-lines: <line number>``
+     - Lines to highlight for the code sample. You can specify a range 
+       of line numbers, individual line numbers, or both.
+
+       **Default:** None
+ 
+   * - ``:caption: <caption text>``
+     - The caption for the code sample.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:380]:
```suggestion
     - Brief description.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 10]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** None
+
+   * - ``:dedent: <number of spaces>``
+     - The number of spaces to remove from the beginning of each line of code.
+       Use this option if the code you want to show is indented in the code file.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:386]:
```suggestion
     - Number of spaces to remove from the beginning of each line of code.
       Use this option if the code you want to show is indented in the code file.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 11]
```diff
@@ -2,20 +2,444 @@
...
+
+       **Default:** None 
+
+Example 1
+~~~~~~~~~
+
+Suppose you want to include samples from the following Python file in a docs page:
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:393]:
```suggestion
Suppose you want to include samples from the following Python file in a documentation page:
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 12]
```diff
@@ -2,20 +2,444 @@
...
+.. literalinclude:: /includes/python-sample.py
+   :copyable: false
+   :end-before: print("...but this line won't.")
+   :caption: Sample Python code from another file
+
+Example 2
+~~~~~~~~~
+
+You can use comments to mark the points where the ``literalinclude`` directive
+starts and finishes pulling in code.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:419]:
```suggestion
starts and finishes including code.
```

File: source/style-guide/markup/directives/code-blocks.txt [hunk 13]
```diff
@@ -2,20 +2,444 @@
...
+By using comments instead of code, you can unambiguously specify the start and end of
+the code sample, regardless of how many samples you're pulling in from the file.
```
Comment (corryroot) [inline source/style-guide/markup/directives/code-blocks.txt:421]:
```suggestion
the code example, regardless of how many examples you're including from the file.
```

**Review by corryroot — CHANGES_REQUESTED** [review]
Thanks @mongoKart! I commented with some suggestions.

File: source/style-guide/markup/directives/code-blocks.txt [hunk 14]
```diff
@@ -2,20 +2,444 @@
 Code Examples
 =============
 
-As needed, use the appropriate language option for code blocks directives
-including the following directives:
+Snooty supports three directives for including code samples in a docs page:
```
[the remark this answers is not in the document]
Comment (lindseymoore) [reply source/style-guide/markup/directives/code-blocks.txt:5]: Agreed! A TOC in the right nave pane would be very helpful.

**Review by lindseymoore — APPROVED** [review]
Pending implementation of the other comments, but LGTM!

File: source/style-guide/markup/directives/code-blocks.txt [hunk 15]
```diff
@@ -2,20 +2,444 @@
...
+      :end-before: <line of code>
+      :linenos:
+      :lineno-start: <line number>
+      :emphasize-lines: <line numbers>
+      :caption: <caption>
+      :dedent: <number>
+
+Options
+~~~~~~~
+
```
Comment (jeff-allen-mongo) [inline source/style-guide/markup/directives/code-blocks.txt:330]: Maybe consider adding a sentence that says all of these options are optional.

File: source/style-guide/markup/directives/code-blocks.txt [hunk 16]
```diff
@@ -2,20 +2,444 @@
...
 
 - ``code-block``
 - ``io-code-block``
 - ``literalinclude``
 
-These directives feature syntax highlighting which makes the source code 
-contained in them easier to interpret. The following example shows how to 
-specify Python language syntax highlighting in a ``code-block`` directive:
+These directives feature a monospace font and syntax highlighting that makes the
+source code contained in them easier to interpret.
```
Comment (jeff-allen-mongo) [inline source/style-guide/markup/directives/code-blocks.txt:12]: [Suggestion for minimalism]

```suggestion
These directives feature a monospace font and syntax highlighting that makes
source code easier to interpret.
```

**Review by jeff-allen-mongo — APPROVED** [review]
LGTM % some non-blocking suggestions. Thanks for this update. I think this will be super useful for writers.

File: source/style-guide/markup/directives/code-blocks.txt [hunk 17]
```diff
@@ -2,20 +2,444 @@
...
+   * - ``<language>``
+     - The language to use for syntax highlighting. 
+       
+       For a complete list of supported languages, see the
+       `leafygreen-ui GitHub repository. <https://github.com/mongodb/leafygreen-ui/tree/main/packages/code#properties>`__
+
+       **Default:** None
+ 
+   * - ``:copyable: <boolean>``
+     -  Whether to show the Copy button for the code sample.
```
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/style-guide/markup/directives/code-blocks.txt:48]: I think it's because it's a div, which lacks an alt text tag. However, I reached out to Design Systems to ask whether they could add text when you hover over it which could be better than nothing.

File: source/style-guide/markup/directives/code-blocks.txt [hunk 18]
```diff
@@ -2,20 +2,444 @@
...
 - ``io-code-block``
 - ``literalinclude``
 
-These directives feature syntax highlighting which makes the source code 
-contained in them easier to interpret. The following example shows how to 
-specify Python language syntax highlighting in a ``code-block`` directive:
+These directives feature a monospace font and syntax highlighting that makes the
+source code contained in them easier to interpret.
+
+Code Blocks
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt:14]: How do you envision this being formatted? With a similar tabbed interface?

File: source/style-guide/markup/directives/code-blocks.txt [hunk 19]
```diff
@@ -2,20 +2,444 @@
...
+
+   * - ``:source: <url>``
+     - The URL of the code file. This option adds a link to the 
+       specified URL on the right side of the code sample.
+       
+       You must also specify ``:copyable: true`` (or omit the ``:copyable:`` option)
+       for this option to work.
+       
+       **Default:** None
+
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: When I was originally working on the ticket, [Cassidy confirmed the attribute is deprecated.](https://mongodb.slack.com/archives/C018UTGP23T/p1698421402661889?thread_ts=1698257615.434079&cid=C018UTGP23T)

Do you still think it should be added?

File: source/style-guide/markup/directives/code-blocks.txt [hunk 20]
```diff
@@ -2,20 +2,444 @@
...
+      html_theme = sconf.theme.name
+      html_theme_path = [os.path.join(conf.paths.buildsystem, 'themes')]
+      html_title = conf.project.title
+      htmlhelp_basename = 'MongoDBdoc'
+
+The previous ``code-block`` directive creates the following output:
+
+.. code-block:: python
+      :source: https://github.com/mongodb/docs-ecosystem/blob/36f9cc7260a246d47cb05ac35276a2c92734b028/conf.py#L120
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: good catch

File: source/style-guide/markup/directives/code-blocks.txt [hunk 21]
```diff
@@ -2,20 +2,444 @@
...
+   :copyable: false
+   :end-before: print("...but this line won't.")
+   :caption: Sample Python code from another file
+
+Example 2
+~~~~~~~~~
+
+You can use comments to mark the points where the ``literalinclude`` directive
+starts and finishes pulling in code.
+By using comments instead of code, you can unambiguously specify the start and end of
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: reworded

File: source/style-guide/markup/directives/code-blocks.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: updated

File: source/style-guide/markup/directives/code-blocks.txt [hunk 22]
```diff
@@ -2,20 +2,444 @@
...
+the code sample, regardless of how many samples you're pulling in from the file.
+You also won't need to update the directive if the extracted code changes.
+
+In the Python file in the previous example, the code inside ``sample_function()`` is
+indented two spaces. The following ``literalinclude`` directive uses the ``:dedent:``
+option to remove these spaces:
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: I'm not sure I understand your explanation; could you clarify?

File: source/style-guide/markup/directives/code-blocks.txt [hunk 23]
```diff
@@ -2,20 +2,444 @@
...
+       **Default:** None
+  
+   * - ``:caption: <caption text>``
+     -  The caption for the code sample.
+
+        **Default:** None 
+
+   * - ``:source: <url>``
+     - The URL of the code file. This option adds a link to the 
+       specified URL on the right side of the code sample.
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: Removed this text to (hopefully) align with a suggestion made by @ccho-mongodb

File: source/style-guide/markup/directives/code-blocks.txt [hunk 24]
```diff
@@ -2,20 +2,444 @@
...
+       **Default:** ``true``
+
+   * - ``:caption: <caption text>``
+     - The caption for the code sample.
+
+       **Default:** None
+
+   * - ``:source: <url>``
+     - The URL of the code file. This option adds a link to the 
+       specified URL on the right side of the code sample.
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: removed

File: source/style-guide/markup/directives/code-blocks.txt [hunk 25]
```diff
@@ -2,20 +2,444 @@
...
+     -  The caption for the code sample.
+
+        **Default:** None 
+
+   * - ``:source: <url>``
+     - The URL of the code file. This option adds a link to the 
+       specified URL on the right side of the code sample.
+       
+       You must also specify ``:copyable: true`` (or omit the ``:copyable:`` option)
+       for this option to work.
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: reworded

File: source/style-guide/markup/directives/code-blocks.txt [hunk 26]
```diff
@@ -2,20 +2,444 @@
...
+   * - ``:caption: <caption text>``
+     - The caption for the code sample.
+
+       **Default:** None
+
+   * - ``:source: <url>``
+     - The URL of the code file. This option adds a link to the 
+       specified URL on the right side of the code sample.
+       
+       You must also specify ``copyable: true`` for this option to work.
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: reworded

File: source/style-guide/markup/directives/code-blocks.txt [hunk 27]
```diff
@@ -2,20 +2,444 @@
...
+      [
+          { title: 'The Life Aquatic with Steve Zissou', ... }
+      ]
+
+Literal Includes
+----------------
+
+The ``literalinclude`` directive formats a code sample pulled in from another file.
+This directive is helpful if you need to keep the code in a separate file
+or if you need to display the same code sample in more than one place.
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: I have no problem with this rewording, but the Style Guide discourages "multiple"

[the remark this answers is not in the document]
Comment (corryroot) [reply source/style-guide/markup/directives/code-blocks.txt]: @mongoKart, can you share a citation? I don't see "multiple" in our list of terms and see that we use it across the Style Guide.

[the remark this answers is not in the document]
Comment (corryroot) [reply source/style-guide/markup/directives/code-blocks.txt]: In this case, I think the [concise terms guidance](https://www.mongodb.com/docs/meta/style-guide/terminology/concise-terms/) provides an argument in favor.

**Review by corryroot — APPROVED** [review]
Thanks @mongoKart! LGTM

File: source/style-guide/markup/directives/code-blocks.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/style-guide/markup/directives/code-blocks.txt]: Following up for @corryroot 
Here's the solution the design systems team decided upon: https://jira.mongodb.org/browse/LG-3874

[the remark this answers is not in the document]
Comment (corryroot) [reply source/style-guide/markup/directives/code-blocks.txt]: @ccho-mongodb, awesome. Thanks!

File: source/reference/code-blocks.txt [hunk 28]
```diff
@@ -0,0 +1,467 @@
...
+
+The previous ``code-block`` directive creates the following output:
+
+.. code-block:: python
+   :source: https://github.com/mongodb/docs-ecosystem/blob/36f9cc7260a246d47cb05ac35276a2c92734b028/conf.py#L120
+
+      html_theme = sconf.theme.name
+      html_theme_path = [os.path.join(conf.paths.buildsystem, 'themes')]
+      html_title = conf.project.title
+      htmlhelp_basename = 'MongoDBdoc'
```
Comment (ccho-mongodb) [inline source/reference/code-blocks.txt:135]: I think for consistency, the attribute indentation should match the column that contains the `:source:` attribute even if the parser accepts this.

File: source/style-guide/markup/directives/code-blocks.txt (hunk 18 above)
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/style-guide/markup/directives/code-blocks.txt]: No changes needed at this time. I was thinking that an example rendered code block could be shown at the top of each section, perhaps in an Example admonition (used by Icon Roles and Image Lightboxes sections). I think this isn't urgent and could be completed in a separate ticket at a future time.

File: source/reference/code-blocks.txt [hunk 29]
```diff
@@ -0,0 +1,467 @@
...
+       extracted code changes.
+
+       **Default:** Beginning of file
+   
+   * - ``:end-before: <line of code>``
+     - Line of code after the last line to include in the sample.
+
+       You can use a comment to mark the point where the ``literalinclude`` directive
+       stops including code. If you do, omit the comment marker (``#`` or ``//``) from
+       the attribute value.
```
Comment (ccho-mongodb) [inline source/reference/code-blocks.txt:381]: Suggestion:

I think based on this statement, it would be helpful to omit the "#" in the final example on the page so that conforms to the guidance. Since omitting is optional, maybe rewording this as "If you do, you can omit..." is the only change needed.

File: source/style-guide/markup/directives/code-blocks.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/style-guide/markup/directives/code-blocks.txt]: I wasn't able to locate this info other than the start-after and end-before list table descriptions. I still think it would be helpful to describe this in the description under the Literal Includes header, but happy to let you decide what you think would work best.

File: source/style-guide/markup/directives/code-blocks.txt (hunk 22 above)
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/style-guide/markup/directives/code-blocks.txt]: I don't remember if I had seen something in the description of the dedent option that needed clarification here, but looks fine to me now.

File: source/reference/code-blocks.txt [hunk 30]
```diff
@@ -0,0 +1,467 @@
...
+
+All of the following options are optional.
+
+.. list-table::
+
+   * - ``<language>``
+     - Language to use for syntax highlighting. 
+       
+       For a complete list of supported languages, see the
+       `leafygreen-ui GitHub repository. <https://github.com/mongodb/leafygreen-ui/tree/main/packages/code#properties>`__
```
Comment (ccho-mongodb) [inline source/reference/code-blocks.txt:50]: Issue:
I think this link shows the code for rendering the HTML of a code component and not the supported languages.

Suggestion:
I think https://github.com/mongodb/leafygreen-ui/blob/main/packages/code/src/languages.ts lists the supported languages.

**Review by ccho-mongodb — APPROVED** [review]
LGTM. Responded to a few comments from the previous review. Identified only minor items, no additional review required.

File: source/style-guide/markup/directives/code-blocks.txt (hunk 27 above)
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/style-guide/markup/directives/code-blocks.txt]: You're right--I don't see it in the Style Guide proper either. Where I do see it is in `TooWordy.yml`, a Vale rule that came with the `mongodb-vale` repo and which I see popups for in VS Code. If these rules don't accurately reflect our style, I would love to get them changed.

Like I said, "multiple" works for me!

File: source/reference/code-blocks.txt (hunk 29 above)
Comment (mongoKart) [reply source/reference/code-blocks.txt:381]: good thought
